Proposal to use unicode and include proofs in some snippets

williamdemeo opened the pull request:
# Description

## Proposals

1. We should freely use unicode.
2. We should also make the snippets type-checkable when it's not too inconvenient to do so; e.g., when it doesn't require too many extra lines or excessive, distracting syntax.
3. We should add at least one literate `lagda` file to the end of each section; this file would collect all of code snippets for that section into a comprehensive, type-checkable Agda program; the file `Section-1-1.lagda` is a small example.

williamdemeo committed:
initial commit of dependencies for agda version and agda config files

+ fixes compile problem
+ makes git status more readable
+ other improvements and additions

williamdemeo commented on src/content/1.1/code/agda/snippet01.agda:
```diff
@@ -1 +1 @@
-f : A -> B
\ No newline at end of file
+f : A → B
```
Do you prefer to avoid unicode?

lemastero replied:
I do not have strong opinion on this.

Pros of ASCII:
1. some Agda experts like Conor McBride promote `{-# OPTIONS --no-unicode #-}` style
2. it might be easier for programmers (who are target audience) to see ASCII arrows -> rather than unicode ones →

Cons:
3. some Agda experts prefer Unicode e.g. Martin Escardo, Philip Wadler, etc.
4. standard library uses Unicode heavily 

I was slightly for ASCII due to 2. But let's go for Unicode - we can always change our mind later.

lemastero replied:
We agreed to go for unicode.

williamdemeo commented on src/content/1.1/code/agda/snippet04.agda:
```diff
@@ -1,4 +1,10 @@
+     (h : C → D)
+  →  (h ∘ g) ∘ f ≡ h ∘ (g ∘ f)
+
+_ = λ f g h → _≡_.refl
```
I think, where possible and where it's not too inconvenient and doesn't require too much extra code, we should try to write snippets that type-check successfully and, where appropriate, include proofs of stated properties (e.g., associativity in this example).  In Haskell and other languages, it's not clear that including formal proofs is a reasonable goal, but in Agda it certainly is. :)

lemastero replied:
Agree on this. OTOH in many places at the beginning they just show syntax - I will re-read this.

williamdemeo commented on src/content/1.1/code/agda/snippet05.agda:
```diff
@@ -1,2 +1,2 @@
 id : ∀ {A} → A → A
-id a = a
\ No newline at end of file
+id a = a
```
I don't know what changed here.

lemastero replied:
Here and in e.g. in other places like `snippet03.agda` you are adding newline at the end of the file. Technically this is correct, but for some case snippets does not contains them.

Normally people insist on having newline at the end of the every text file. Otherwise it is not technically text file. That is why GH shows red circle below old version. To note newline is missing.

We can go for newlines and if we have PDF version working test if there is any reason to remove them. It should be easy to automate.

williamdemeo commented on src/content/2.1/code/agda/snippet01.lagda:
```diff
@@ -0,0 +1,12 @@
+      → (∀{x}(y : B x) → f y) → (g : (x : A) → B x)
+      → ((x : A) → f (g x))
+f ∘ᵈ g = λ x → f (g x)
+\end{code}
```
oops, I didn't mean to include this file.  I'll remove it from the PR.

williamdemeo commented on src/content/agda/editor-note.tex:
```diff
@@ -0,0 +1,38 @@
+\optlangcode
+\NoIndentAfterThis
+% In addition, some Scala snippets make use of the
+% \urlref{https://github.com/non/kind-projector}{Kind Projector} compiler plugin, to support nicer syntax for partially-applied types.
```
This file is required for compiling the agda version of ctfp.

williamdemeo commented on src/ctfp-print-agda.tex:
```diff
@@ -0,0 +1,3 @@
+\input{opt-agda}
+\input{ctfp-print}
+\input{postamble}
```
I believe this and the remaining files of the PR are also required for compiling the agda version of ctfp (with the command `nix build .#ctfp-agda`)

williamdemeo commented on src/preamble.tex:
```diff
@@ -76,7 +76,8 @@
 
-\usepackage[cachedir=_minted-ctfp]{minted}
+%\usepackage[cachedir=_minted-ctfp]{minted}
+\usepackage[cache=false]{minted}
```
For some reason, I needed this mod in order for ctfp to build on my machine.

williamdemeo commented on src/preamble.tex:
```diff
@@ -144,6 +145,8 @@
 \usepackage{xparse}
 \usepackage[abspath]{currfile}
 
+\usepackage{xspace}
```
I need this for some of the latex commands I defined.

lemastero reviewed (approved):
Great job!
`nix build .#ctfp-agda` worked for me
I have tried to add support for Agda [on my own](https://github.com/lemastero/milewski-ctfp-pdf/pull/2/files), but I wan not able to fix build.

lemastero commented on src/content/1.1/code/agda/snippet04.agda:
```diff
@@ -1,4 +1,10 @@
-h : C -> D
-h ∘ (g ∘ f) == (h ∘ g) ∘ f == h ∘ g ∘ f
\ No newline at end of file
+open import Function using (_∘_)
```
![image](https://github.com/formalverification/milewski-ctfp-pdf/assets/10901543/ff6647f2-b573-4739-bcc6-7ca116d1e1d4)

Function composition does not render properly for me.

williamdemeo replied:
This is a really annoying problem that crops up sometimes for me too.  It has to do with not using a font that includes that symbol.  I’m not sure how to fix it.  Usually when I fix such a problem, my solution is ad-hoc or by trial and error, so I can never remember how I solved it… but this would be a good thing to address more abstractly so we can document it for future cases where the problem arises.

lemastero commented on src/content/1.1/code/agda/snippet04.agda:
```diff
@@ -1,4 +1,10 @@
+     (h : C → D)
+  →  (h ∘ g) ∘ f ≡ h ∘ (g ∘ f)
+
+_ = λ f g h → _≡_.refl
```
I wonder if explicitly importing refl would be a bit easier for readers:
```agda
open import Function using (_∘_)
open import Relation.Binary.PropositionalEquality using (_≡_; refl)

_ :  {A B C D : Set}
     (f : A → B)
     (g : B → C)
     (h : C → D)
  →  (h ∘ g) ∘ f ≡ h ∘ (g ∘ f)

_ = λ f g h → refl
```

williamdemeo replied:
Yes, let's change to that convention (i.e., we should import refl).  By the way, it appears the composition operator is working for you now... yes?

williamdemeo committed:
add .orig to .gitignore to make `git status` output cleaner

williamdemeo commented:
It seems the Prettier check doesn't like the flake.lock... but I'm not sure we should have the flake.lock in the repository... ?  I'll try removing it and see if that doesn't break something.

williamdemeo committed:
remove flake.lock

williamdemeo committed:
mention Agda in README

Piotr did this already and it might be what's causing the Prettier
test to faile.  We'll know for sure if Prettier fails when I push
this change.

williamdemeo committed:
fix long line so Prettier doesn't complain

williamdemeo committed:
remove extra white space from the end of line 36

to satisfy Prettier tests
